backport-2.1: engine: fix bug in Distinct() after ApplyBatchRepr()

tbg opened the pull request:
Backport 1/1 commits from #34493.

/cc @cockroachdb/release

---

Release note: None

cockroach-teamcity commented:
<!-- Reviewable:start -->
This change is [<img src="https://reviewable.io/review_button.png" height="34" align="absmiddle" alt="Reviewable"/>](https://reviewable.io/reviews/cockroachdb/cockroach/34773)
<!-- Reviewable:end -->

tbg committed:
engine: fix bug in Distinct() after ApplyBatchRepr()

We weren't marking a Distinct batch as needing a flush after the
parent batch applied a WriteBatch, so the Distinct batch would
not observe the WriteBatch.

I audited the code and it seems that nothing is affected by it.
The most dangerous code path is here:

https://github.com/cockroachdb/cockroach/blob/1b8689c0b4df102e1bf4e271913c4bb096ca8ffe/pkg/storage/replica_raft.go#L2201

But the current code appears to never read from the batch (I found the
bug there because I was introducing code that *did* read from it).

Release note: None

petermattis reviewed (approved):
<img class="emoji" title=":lgtm:" alt=":lgtm:" align="absmiddle" src="https://reviewable.io/lgtm.png" height="20" width="61"/>

> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/34773#-:-LYRz0a66OP32XW0f2ve:bnfp4nl)* status: :shipit: complete! 1 of 0 LGTMs obtained (waiting on @petermattis)



<!-- Sent from Reviewable.io -->

bdarnell commented:
LGTM. Does this need a release note or do we think this bug wasn't causing a problem before in 2.1?

tbg commented:
I did a light audit of it and it didn't seem that there was an obvious bug in 2.1 related to this.
